firedrake/functionspacedata.py

```diff
@@ -175,32 +181,77 @@ def get_dof_offset(mesh, key, entity_dofs, ndof):
+        point p (for "topological", ndof basis functions with non-zero
+        support on points in the closure of p for "geometric").  The
+        basis function indices are in the index array, starting at
+        section.getOffset(p).
     """
-    if not bool(mesh.layers):
+    if not mesh.cell_set._extruded:
         return None
-    bt_masks = {}
-    # Compute the top and bottom masks to identify boundary dofs
-    #
-    # Sorting the keys of the closure entity dofs, the whole cell
-    # comes last [-1], before that the horizontal facet [-2], before
-    # that vertical facets [-3]. We need the horizontal facets here.
-    closure_dofs = finat_element.entity_closure_dofs()
-    horiz_facet_dim = sorted(closure_dofs.keys())[-2]
-    b_mask = closure_dofs[horiz_facet_dim][0]
-    t_mask = closure_dofs[horiz_facet_dim][1]
-    bt_masks["topological"] = (b_mask, t_mask)  # conversion to tuple
-    # Geometric facet dofs
-    facet_dofs = entity_support_dofs(finat_element, horiz_facet_dim)
-    bt_masks["geometric"] = (facet_dofs[0], facet_dofs[1])
-    return bt_masks
+    masks = {}
+    _, kind = key
+    assert kind in {"cell", "interior_facet"}
+    dim = finat_element.cell.get_spatial_dimension()
+    ecd = finat_element.entity_closure_dofs()
+    try:
+        esd = finat_element.entity_support_dofs()
+    except NotImplementedError:
+        # 4-D cells
+        esd = None
+    # Number of entities on cell excepting the cell itself.
+    chart = sum(map(len, ecd.values())) - 1
+    closure_section = PETSc.Section().create(comm=PETSc.COMM_SELF)
+    support_section = PETSc.Section().create(comm=PETSc.COMM_SELF)
+    # Double up for interior facets.
+    if kind == "cell":
+        ncell = 1
+    else:
+        ncell = 2
+    closure_section.setChart(0, ncell*chart)
+    support_section.setChart(0, ncell*chart)
+    closure_indices = []
+    support_indices = []
+    facet_points = []
+    p = 0
+
+    offset = finat_element.space_dimension()
+    for cell in range(ncell):
+        for ent in sorted(ecd.keys()):
+            # Never need closure of cell
+            if sum(ent) == dim:
+                continue
+            for key in sorted(ecd[ent].keys()):
+                closure_section.setDof(p, len(ecd[ent][key]))
+                vals = numpy.asarray(sorted(ecd[ent][key]), dtype=IntType)
+                closure_indices.extend(vals + cell*offset)
+                if esd is not None:
+                    support_section.setDof(p, ncell*len(esd[ent][key]))
+                    vals = numpy.asarray(sorted(esd[ent][key]), dtype=IntType)
+                    support_indices.extend(vals + cell*offset)
+                if sum(ent) == dim - 1:
+                    facet_points.append(p)
+                p += 1
+    closure_section.setUp()
+    support_section.setUp()
+    closure_indices = numpy.asarray(closure_indices, dtype=IntType)
+    support_indices = numpy.asarray(support_indices, dtype=IntType)
```

Comment:
So these are CSR arrays that index which dofs are on the boundary.  The CSR encoding is in the section, `section.getDof(p) > 0` says that dofs on entity `p` are on the boundary, the offset into the array is given by `section.getOffset(p)`.